src/gc/impl/manual/gc.d

```diff
@@ -0,0 +1,216 @@
+
+        pthis.gc_addRoot = function void(void* p) { gcInstance.addRoot(p);};
+        pthis.gc_addRange = function void(void* p, size_t sz, const TypeInfo ti) { gcInstance.addRange( p, sz, ti );};
+
+        pthis.gc_removeRoot = function void(void* p) { gcInstance.removeRoot(p);};
+        pthis.gc_removeRange = function void(void*p) { gcInstance.removeRange(p);};
+        pthis.gc_runFinalizers = function void(in void[] segment) { };
+
+        pthis.gc_inFinalizer = function bool() { return false;};
+
+        pthis.gc_setProxy = function void(Proxy* p) {
+            foreach( r; gcInstance.roots[0 .. gcInstance.nroots] )
+                p.gc_addRoot( r );
+
+            foreach( r; gcInstance.ranges[0 .. gcInstance.nranges] )
+                p.gc_addRange( r.pos, r.len, r.ti );
+        };
+        pthis.gc_clrProxy = function void(Proxy* p){
+            foreach( r; gcInstance.ranges[0 .. gcInstance.nranges] )
+                p.gc_removeRange( r.pos );
+
+            foreach( r; gcInstance.roots[0 .. gcInstance.nroots] )
+                p.gc_removeRoot( r );
+        };
+    }
+
+    void Dtor()
+    {
+        free(roots);
+        free(ranges);
+    }
+
+    void* malloc(size_t sz) nothrow
+    {
+        void* p = cstdlib.malloc( sz );
+
+        if( sz && p is null )
+            onOutOfMemoryError();
+        return p;
+    }
+
+    void* calloc( size_t sz ) nothrow
+    {
+        void* p = cstdlib.calloc( 1, sz );
+
+        if( sz && p is null )
+            onOutOfMemoryError();
+        return p;
+    }
+
+    void* realloc( void* p, size_t sz ) nothrow
+    {
+        p = cstdlib.realloc( p, sz );
+
+        if( sz && p is null )
+            onOutOfMemoryError();
+        return p;
+    }
+
+    void free(void* p) nothrow
+    {
+        cstdlib.free(p);
+    }
+
+    void addRoot(void* p) nothrow
+    {
+        void** r = cast(void**) realloc( roots, (nroots+1) * roots[0].sizeof );
+        if( r is null )
+            onOutOfMemoryError();
+        r[nroots++] = p;
+        roots = r;
+    }
+
+    void addRange(void* p, size_t sz, const TypeInfo ti = null) nothrow
+    {
+        Range* r = cast(Range*) realloc( ranges, (nranges+1) * ranges[0].sizeof );
+        if( r is null )
+            onOutOfMemoryError();
+        r[nranges].pos = p;
+        r[nranges].len = sz;
+        r[nranges].ti = cast()ti;
+        ranges = r;
+        ++nranges;
+    }
+
+    void removeRoot(void* p) nothrow
+    {
+        for( size_t i = 0; i < nroots; ++i )
+        {
+            if( roots[i] is p )
+            {
+                roots[i] = roots[--nroots];
+                return;
+            }
+        }
+        assert( false );
+    }
+
+    void removeRange(void *p) nothrow
+    {
+        for( size_t i = 0; i < nranges; ++i )
+        {
+            if( ranges[i].pos is p )
+            {
+                ranges[i] = ranges[--nranges];
+                return;
+            }
+        }
+        assert( false );
+    }
+}
+
+
```

Comment:
For new code, please use [dfmt](https://github.com/Hackerpilot/dfmt) `dfmt -i file.d` and make sure to read the [The D Style Guide](https://dlang.org/dstyle.html).